### PR TITLE
CSP-554-GetProfilesEndPoint

### PR DESCRIPTION
**Pull request by YuvrajPatil-DFE**
Implemented GetProfiles endpoint and test cases for the same.

### CODE REVIEW

File: src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj [hunk 1]
```diff
@@ -8,6 +8,10 @@
     <IsPackable>false</IsPackable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="Profiles\Queries\GetProfilesQueryResultTests.cs" />
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj:12]: Not sure what this is about?

File: src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQuery.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using MediatR;
+using SFA.DAS.AANHub.Application.Mediatr.Responses;
+
+namespace SFA.DAS.AANHub.Application.Profiles.Queries.GetProfiles
+{
+    public class GetProfilesQuery : IRequest<ValidatedResponse<List<ProfileModel>>>
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQuery.cs:6]: Rename this and all other respective files including tests, to use `GetProfilesByUserType`

File: src/SFA.DAS.AANHub.Domain/Interfaces/Repositories/IProfilesReadRepository.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using SFA.DAS.AANHub.Domain.Entities;
+
+namespace SFA.DAS.AANHub.Domain.Interfaces.Repositories
+{
+    public interface IProfilesReadRepository
+    {
+        Task<List<Profile>> GetAllProfiles(string userType);
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Domain/Interfaces/Repositories/IProfilesReadRepository.cs:7]: GetProfilesByUserType

File: src/SFA.DAS.AANHub.Application/SFA.DAS.AANHub.Application.csproj [hunk 4]
```diff
@@ -6,6 +6,10 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="Profiles\Queries\GetProfiles\GetProfilesQueryResult.cs" />
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/SFA.DAS.AANHub.Application.csproj:10]: Delete the result file, you have excluded it, hence it is in the csproj and also the file is in the PR

File: src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryResult.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿namespace SFA.DAS.AANHub.Application.Profiles.Queries.GetProfiles
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryResult.cs:1]: Delete this file

File: src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj (hunk 1 above)
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj:12]: You need to delete the test file which is not required.

File: src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/GetProfilesQueryResultTests.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿namespace SFA.DAS.AANHub.Application.UnitTests.Profiles.Queries
+{
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/GetProfilesQueryResultTests.cs:2]: Delete this file

File: src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/ProfileModelTests.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace SFA.DAS.AANHub.Application.UnitTests.Profiles.Queries
+{
+    [TestFixture]
+    public class ProfileModelTests
+    {
+        [Test]
+        public void ProfileModelTest_ReturnsExpectedFields()
+        {
+            const string category = "Events";
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/ProfileModelTests.cs:14]: Could use AutoData attribute to generate random values for Profile

File: src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        /// <summary>
+        /// Get list of profiles
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(List<ProfileModel>), StatusCodes.Status200OK)]
+        public async Task<IActionResult> GetProfiles([FromQuery] string userType)
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs:27]: GetProfilesByUserType

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+using SFA.DAS.AANHub.Application.UnitTests;
+
+namespace SFA.DAS.AANHub.Api.UnitTests.Controllers
+{
+    public class ProfilesControllerTests
+    {
+        private readonly ProfilesController _controller;
+        private readonly Mock<IMediator> _mediator;
+
+        public ProfilesControllerTests()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs:21]: Delete this unused ctor and class level variables

**Review by DevenKShah — COMMENTED** [review]
Minor suggestions

File: src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj (hunk 1 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj:12]: Deleted.

File: src/SFA.DAS.AANHub.Domain/Interfaces/Repositories/IProfilesReadRepository.cs (hunk 3 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Domain/Interfaces/Repositories/IProfilesReadRepository.cs:7]: Done

File: src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/ProfileModelTests.cs (hunk 7 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/ProfileModelTests.cs:14]: Implemented

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+            mediatorMock.Setup(m => m.Send(It.Is<GetProfilesByUserType>(q => q.UserType == userType), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(response);
+
+            var getResult = await sut.GetProfilesByUserType(userType);
+
+            var result = getResult as OkObjectResult;
+            result.Should().NotBeNull();
+            Assert.AreEqual(StatusCodes.Status200OK, result!.StatusCode);
+        }
```
Comment (ravihejmadi) [inline src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs:37]: We could add a few more tests for NotFound. There are a few examples in the ControllerTests for the other entities - i.e. Apprentice within ApprenticesControllerTests.cs

File: src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IMediator _mediator;
+        public ProfilesController(ILogger<ProfilesController> logger, IMediator mediator)
+        {
+            _logger = logger;
+            _mediator = mediator;
+        }
+
+        /// <summary>
+        /// Get list of profiles by user type
+        /// </summary>
```
Comment (ravihejmadi) [inline src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs:22]: Please add 
`        /// <param name="userType"></param>`

Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs:22]: Done.

File: src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace SFA.DAS.AANHub.Application.Profiles.Queries.GetProfiles
+{
+    public class GetProfilesQueryValidator : AbstractValidator<GetProfilesByUserType>
+    {
+        public GetProfilesQueryValidator()
+        {
+            RuleFor(a => a.UserType)
+                .NotEmpty()
+                .IsEnumName(typeof(UserType), caseSensitive: false);
```
Comment (ravihejmadi) [inline src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs:12]: Do we need a test for this? I can see that the UserType None is not included in the TestCase

**Review by ravihejmadi — COMMENTED** [review]
@YuvrajPatil-DFE, couple of minor changes. Re the validatortest recommendation - Deven can take a decision :-)

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs (hunk 10 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Api.UnitTests/Controllers/ProfilesControllerTests.cs:37]: Implemented

File: src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        /// <returns></returns>
+        [HttpGet]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(List<ProfileModel>), StatusCodes.Status200OK)]
+        public async Task<IActionResult> GetProfilesByUserType([FromQuery] string userType)
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Api/Controllers/ProfilesController.cs:28]: Its from ASP.NETCore.MVC and it specifies that a parameter or property should be bound using the request query string.

File: src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs (hunk 12 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs]: I have added testcase for "None" usertype in GetProfilesQueryValidatorTests however, need to check with others if we want to implement it this way or validate it in QueryValidator to avoid executing further for this usertype

File: src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/GetProfilesQueryValidatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        [TestCase("apprenticE", true)]
+        [TestCase("Employer", true)]
+        [TestCase("PARTNER", true)]
+        [TestCase("admin", true)]
+        [TestCase("None", true)]
+        [TestCase(null, false)]
+        [TestCase("", false)]
+        [TestCase(" ", false)]
+        [TestCase("xyz", false)]
```
Comment (MarkFCain) [inline src/SFA.DAS.AANHub.Application.UnitTests/Profiles/Queries/GetProfilesQueryValidatorTests.cs:18]: |I don't understand why the test for 'xyz' is false :)

File: src/SFA.DAS.AANHub.Database/Tables/Profile.sql [hunk 15]
```diff
@@ -4,5 +4,5 @@
     [UserType] NVARCHAR(20) NOT NULL ,
     [Category] NVARCHAR(20) NOT NULL, 
     [Description] NVARCHAR(200) NOT NULL, 
-    [Ordering] TINYINT NOT NULL
+    [Ordering] INT NOT NULL
```
Comment (MarkFCain) [inline src/SFA.DAS.AANHub.Database/Tables/Profile.sql:7]: why has this changed?

**Review by MarkFCain — APPROVED** [review]
just a comment to ignore about a test, and query on why a change to a table field type

File: src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs (hunk 12 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application/Profiles/Queries/GetProfiles/GetProfilesQueryValidator.cs]: I am happy for this PR to proceed to Test. We can look at any additional tests later

**Review by ravihejmadi — APPROVED** [review]

**Review by ravihejmadi — APPROVED** [review]
